src/storage/compression/bitpacking.cpp

```diff
@@ -949,16 +966,37 @@ void BitpackingSkip(ColumnSegment &segment, ColumnScanState &state, idx_t skip_c
 	scan_state.Skip(segment, skip_count);
 }
 
+template <class T>
+unique_ptr<CompressedSegmentState> BitpackingInitSegment(ColumnSegment &segment, block_id_t block_id,
+                                                         optional_ptr<ColumnSegmentState> segment_state) {
+	// segment_state is not used for Bitpacking (no specialized serialize/deserialize methods)
+	// block_id would equal INVALID_BLOCK for a constant value block (not materialized)
+	// auto &buffer_manager = BufferManager::GetBufferManager(segment.db);
+	if (block_id == INVALID_BLOCK) {
```

Comment:
I'm not sure this is correct
InitSegment is called during compression when a segment is created to be written to

Constant compression happens during flushing of the segment. If the stats created during compression indicate that the segment is constant.

If block_id is INVALID_BLOCK I imagine this means that the segment is transient, not connected to a physical block yet (which will happen during flushing)

Reply:
That sounds plausible, yes. (But note the call is made via `compression_init_segment` and not `init_compression`.)

I probably misunderstood parts of the compression logic, it touches a lot of the core code and the indirect ion is beautiful but hard to grasp from reading the code alone :-).

I looked at https://github.com/duckdb/duckdb/blob/4488c61ee780635e67abe1b6164f2cdfadc21b65/src/storage/compression/string_uncompressed.cpp#L197 for an idea of how this works, but clearly overlooked something as the code only works when you reopen the database (a behaviour that could be in line with the idea that this is related to the segment still being "transient" when checkpoint is called?)

Reply:
> this is related to the segment still being "transient" when checkpoint is called

Close, but just to clarify:
`block_id` is equal to INVALID_BLOCK when the segment is first created, as part of the compression code.
When checkpointing, the existing data+updates are scanned, this is the `input` of the compress function. In the compress function we create a new segment to write the compressed data to.
This new segment is what InitSegment gets called on.
Before any data has been written to the segment